lib/versioner/staggered.go

```diff
@@ -309,3 +302,65 @@ func (v *Staggered) Archive(filePath string) error {
 
 	return nil
 }
+
+func renameOrCopy(from fs.Filesystem, fromPath string, to fs.Filesystem, toPath string) (err error) {
+	defer from.Remove(fromPath)
+
+	if from.Type() == to.Type() {
+		fromUri := from.URI()
+		toUri := to.URI()
+		// See if one is a prefix of the other filesystem.
+		if strings.HasPrefix(toUri, fromUri) {
```

Comment:
I suspect this suffers from the usual path prefix vulnerability (`c:\foo` being a prefix of `c:\foobar` but actually unrelated). I also (still) think this whole thing is overthinking it. To retain the previous behavior we can just `if from.Type() == fsTypeBasic && from.Type() == to.Type() { return os.Rename(...) }`. For non-basic filesystems (hypothetical at this point) there is no reason to assume that a shared prefix means a rename is possible. I don't think that's the case between S3 buckets for example, even though they may share a region or whatever as prefix.

So all in all I think we take damage from trying to be clever here. This is my only real complaint and that I would want adressed before merging. Otherwise it looks good to me and I think the proper course of action is to merge and get it out there in a candidate for some real world testing.